gpiozero/pins/__init__.py

```diff
@@ -6,39 +8,133 @@
     )
 str = type('')
 
-import io
-
-from .data import pi_info
 from ..exc import (
     PinInvalidFunction,
     PinSetInput,
     PinFixedPull,
+    PinSPIUnsupported,
     PinPWMUnsupported,
     PinEdgeDetectUnsupported,
+    SPIFixedClockMode,
+    SPIFixedBitOrder,
+    SPIFixedSelect,
+    SPIFixedWordSize,
     )
 
 
-PINS_CLEANUP = []
-def _pins_shutdown():
-    for routine in PINS_CLEANUP:
-        routine()
+class Factory(object):
+    """
+    Generates pins, SPI, and I2C interfaces for devices. This is an abstract
+    base class for pin factories. Descendents must override:
+
+    * :meth:`_get_address`
+    * :meth:`pin_address`
+
+    Descendents may override:
+
+    * :meth:`close`
+    * :meth:`pin`
+    * :meth:`spi`
+    * :meth:`_get_pi_info`
+    """
+
+    def close(self):
+        """
+        Closes the pin factory. This is expected to clean up all resources
+        manipulated by the factory. It it typically called at script
+        termination.
+        """
+        pass
+
+    def pin(self, spec):
+        """
+        Creates an instance of a :class:`Pin` descendent representing the
+        specified pin.
+
+        .. warning::
+
+            Descendents must ensure that pin instances representing the same
+            hardware are identical; i.e. two separate invocations of
+            :meth:`pin` for the same pin specification must return the same
+            object.
+        """
+        raise PinGPIOUnsupported("GPIO not supported by this pin factory")
```

Comment:
It's an interesting idea; like a "capabilities" property that lists whether the pin can handle PWM output, analog-in, etc. It's a decent idea but I think one that can marked "future work" for now (open a ticket for it?)